Bug when quitting (q) after running something

After running some executables, when quitting, there is an error message and the OS crashes:
```
panic: Fatal exception 4 (Address error on load) in kernel mode
panic: EPC 0x80005864, exception vaddr 0xdeadbf07
panic: I can't handle this... I think I'll just die now...
```

This is most likely due to the filetable_cleanup function which does nasty things with pointers. Remember to fix.

--- a/kern/fs/filetable.c
+++ b/kern/fs/filetable.c
@@ -46,20 +46,27 @@ filetable_init(void)
     int err_in, err_out, err_err;
     char con_filename[5];
 
+    /* initialize system filetable lock */
     sys_filetable.lock = lock_create("sys_filetable_lock");
     if (sys_filetable.lock == NULL) {
         return ENOMEM;
     }
 
+    /* create 3 file structs for standard files */
     sys_filetable.stdin = (struct fs_file *) kmalloc(sizeof(struct fs_file));
     sys_filetable.stdout = (struct fs_file *) kmalloc(sizeof(struct fs_file));
     sys_filetable.stderr = (struct fs_file *) kmalloc(sizeof(struct fs_file));
 
+    /* if unable to allocate files */
     if (sys_filetable.stdin == NULL || sys_filetable.stdout == NULL
      || sys_filetable.stderr == NULL) {
         return ENOMEM;
      }
 
+    /*
+     * open standard files attached to "con:", need a strcpy of the const string
+     * because vfs_open destroys the filename
+    */
     strcpy(con_filename, "con:");
     err_in = vfs_open(con_filename, O_RDONLY, 0644, &sys_filetable.stdin->f_vnode);
     strcpy(con_filename, "con:");
@@ -67,6 +74,7 @@ filetable_init(void)
     strcpy(con_filename, "con:");
     err_err = vfs_open(con_filename, O_WRONLY, 0644, &sys_filetable.stderr->f_vnode);
 
+    /* check for errors during opening */
     if (err_in) {
         return err_in;
     }
@@ -79,6 +87,7 @@ filetable_init(void)
         return err_err;
     }
 
+    /* create a lock for each file */
     sys_filetable.stdin->f_lock = lock_create("stdin");
     sys_filetable.stdout->f_lock = lock_create("stdout");
     sys_filetable.stderr->f_lock = lock_create("stderr");
@@ -88,6 +97,12 @@ filetable_init(void)
         return ENOMEM;
      }
 
+    /* initialize refcount */
+     sys_filetable.stdin->f_refcount = 1;
+     sys_filetable.stdout->f_refcount = 1;
+     sys_filetable.stderr->f_refcount = 1;
+
+    /* set head and tail to stdin, then add two new files (stdout, stderr) */
     sys_filetable.head = sys_filetable.stdin;
     sys_filetable.tail = sys_filetable.head;
     sys_filetable.head->f_next = NULL;
@@ -95,6 +110,7 @@ filetable_init(void)
     filetable_addfile(sys_filetable.stdout);
     filetable_addfile(sys_filetable.stderr);
 
+    /* init system filetable size */
     sys_filetable.size = 3;
 
     return 0;
@@ -103,9 +119,11 @@ filetable_init(void)
 void
 filetable_cleanup(void)
 {
+    /* destroy system filetable lock */
     lock_destroy(sys_filetable.lock);
     struct fs_file *ptr = filetable_head();
 
+    /* remove every node */
     while (ptr->f_prev != NULL) {
         ptr = ptr->f_prev;
         kfree(ptr->f_next);
@@ -145,18 +163,20 @@ filetable_removefile(struct fs_file *rmfile)
     */
     KASSERT(!(rmfile->f_prev == NULL && rmfile->f_next == NULL));
 
-    /* if prev file is null then this is the tail, so can't edit prev ptr */
+    /* if prev file is not null, we are NOT removing the tail */
     if (rmfile->f_prev != NULL) {
         rmfile->f_prev->f_next = rmfile->f_next;
     } else {
+        /* we are removing the tail, need to move tail pointer */
         sys_filetable.tail = rmfile->f_next;
         sys_filetable.tail->f_prev = NULL;
     }
 
-    /* if next file is null then this is the head, so can't edit next ptr */
+    /* if next file is not null, we are NOT removing the head */
     if (rmfile->f_next != NULL) {
         rmfile->f_next->f_prev = rmfile->f_prev;
     } else {
+        /* we are removing the head, need to move head pointer */
         sys_filetable.head = rmfile->f_prev;
         sys_filetable.head->f_next = NULL;
     }

--- a/kern/fs/filetable.c
+++ b/kern/fs/filetable.c
@@ -66,7 +66,7 @@ filetable_init(void)
     err_out = vfs_open(con_filename, O_WRONLY, 0644, &sys_filetable.stdout->f_vnode);
     strcpy(con_filename, "con:");
     err_err = vfs_open(con_filename, O_WRONLY, 0644, &sys_filetable.stderr->f_vnode);
-    
+
     if (err_in) {
         return err_in;
     }
@@ -137,26 +137,32 @@ void
 filetable_removefile(struct fs_file *rmfile)
 {
     lock_acquire(sys_filetable.lock);
+
+    /*
+    *  if this is the only file in the filetable and you want to close it,
+    *  then there are no stdin, stdout and stderr open so this is going to trigger
+    *  a panic
+    */
+    KASSERT(!(rmfile->f_prev == NULL && rmfile->f_next == NULL));
+
     /* if prev file is null then this is the tail, so can't edit prev ptr */
     if (rmfile->f_prev != NULL) {
         rmfile->f_prev->f_next = rmfile->f_next;
+    } else {
+        sys_filetable.tail = rmfile->f_next;
+        sys_filetable.tail->f_prev = NULL;
     }
 
     /* if next file is null then this is the head, so can't edit next ptr */
     if (rmfile->f_next != NULL) {
         rmfile->f_next->f_prev = rmfile->f_prev;
+    } else {
+        sys_filetable.head = rmfile->f_prev;
+        sys_filetable.head->f_next = NULL;
     }
 
     sys_filetable.size--;
 
-    /*
-    *  if this is the only file in the filetable, make sys_filetable point to
-    *  NULL so that when you add a new file the sys_filetable pointer is reused
-    */
-    if (rmfile->f_prev == NULL && rmfile->f_next == NULL) {
-        sys_filetable.head = NULL;
-        sys_filetable.tail = NULL;
-    }
     /* finally, deallocate the fs_file structure from kernel space */
     lock_destroy(rmfile->f_lock);
     kfree((void *) rmfile);
@@ -164,7 +170,7 @@ filetable_removefile(struct fs_file *rmfile)
 }
 
 /*
-* Returns the size ofstdin the system file pointer
+* Returns the size of the system file pointer
 */
 size_t
 filetable_size(void)
@@ -190,6 +196,9 @@ filetable_tail(void)
     return sys_filetable.tail;
 }
 
+/*
+* Returns pointer to system filetable lock
+*/
 struct lock *
 filetable_lock(void)
 {

--- a/kern/proc/proc.c
+++ b/kern/proc/proc.c
@@ -85,11 +85,6 @@ proc_create(const char *name)
 	/* VFS fields */
 	proc->p_cwd = NULL;
 
-	/* link stdin, stdout and stderr to file descriptors 0, 1, 2 */
-	proc->p_filetable[STDIN_FILENO] = sys_filetable.stdin;
-	proc->p_filetable[STDOUT_FILENO] = sys_filetable.stdout;
-	proc->p_filetable[STDERR_FILENO] = sys_filetable.stderr;
-
 	/* zero all other entries in file descriptor table */
 	for (fd = STDERR_FILENO + 1; fd < OPEN_MAX; fd++) {
 		proc->p_filetable[fd] = NULL;
@@ -101,7 +96,7 @@ proc_create(const char *name)
 
 	/* Exit status initialization */
 	proc->p_exit_status = 0;
-	
+
 	return proc;
 }
 

--- a/kern/proc/proc.c
+++ b/kern/proc/proc.c
@@ -85,6 +85,11 @@ proc_create(const char *name)
 	/* VFS fields */
 	proc->p_cwd = NULL;
 
+	/* link stdin, stdout and stderr to file descriptors 0, 1, 2 */
+	proc->p_filetable[STDIN_FILENO] = sys_filetable.stdin;
+	proc->p_filetable[STDOUT_FILENO] = sys_filetable.stdout;
+	proc->p_filetable[STDERR_FILENO] = sys_filetable.stderr;
+
 	/* zero all other entries in file descriptor table */
 	for (fd = STDERR_FILENO + 1; fd < OPEN_MAX; fd++) {
 		proc->p_filetable[fd] = NULL;

--- a/kern/syscall/runprogram.c
+++ b/kern/syscall/runprogram.c
@@ -97,6 +97,14 @@ runprogram(char *progname)
 		return result;
 	}
 
+	/*
+	* increase the refcount for the stdfiles so that they always have >= 1
+	* refcount and they are not deleted from the sys filetable
+	*/
+	curproc->p_filetable[STDIN_FILENO]->f_refcount++;
+	curproc->p_filetable[STDOUT_FILENO]->f_refcount++;
+	curproc->p_filetable[STDERR_FILENO]->f_refcount++;
+
 	/* Warp to user mode. */
 	enter_new_process(0 /*argc*/, NULL /*userspace addr of argv*/,
 			  NULL /*userspace addr of environment*/,

